test/fail_compilation/b20437.d

```diff
@@ -0,0 +1,14 @@
+/* TEST_OUTPUT:
+---
+fail_compilation/b20437.d(12): Error: cannot implicitly convert expression `d` of type `void delegate()` to `immutable(void delegate())`
+fail_compilation/b20437.d(13): Error: cannot implicitly convert expression `d` of type `void delegate()` to `shared(void delegate())`
+---
+*/
+// https://issues.dlang.org/show_bug.cgi?id=20437
+void f()
+{
+    int x;
+    void delegate() d = delegate() { x = 1; };
```

Comment:
`const` should also be represented here.
`x = 1` is not `const`... if the delegate is const, then how is it that it can perform mutation?

Reply:
I suppose the solution to that is that it should not be possible to call non-const opCall on a const delegate.

```d
const(void delegate()) f = delegate() {}; // ok
f(); // not ok

const(void delegate() const) g = delegate() const {}; // ok
g(); // ok
```

This also seems uncaught by the compiler.

Reply:
What I’m aiming for is that the following types behave the same when it comes to qualifiers and implicit conversions:

```
interface D { R opCall(A...) «qualifiers1»; }
«qualifiers2»(D)

«qualifiers2»(R delegate(A...) «qualifiers1»)
```

I’m not familiar with the exact representation of delegates, so the runtime representation may be very different, but for type checking this seems reasonable.

Reply:
The runtime representation is something like `struct { void* /*context*/ptr; T function(A, void* context) functptr; }` (the argument order is possibly wrong and the ABI is different but the memory layout is as shown).

Reply:
In addition the following conversions should be possible:

```d
void delegate() const     -> const(void delegate() const)
void delegate() immutable -> immutable(void delegate() immutable)
void delegate() shared    -> shared(void delegate() shared)
```

because delegate literals enforce that the variables in their environment are compatible with the qualifiers of opCall, e.g. the following is already prohibited:

```d
int x = 1;
delegate() shared { print(x); } // error: x must be shared
delegate() immutable { print(x); } // error: x must be immutable
```

Reply:
> I suppose the solution to that is that it should not be possible to call non-const opCall on a const delegate.

How do you enforce this though ? If you pass the delegate to a function, you simply can't enforce it (because the function might be `extern`, for example).

Does the qualifier applies to the context, or the delegate itself ?

If it applies to the context, then the usual covariance rules apply (mutable => `const` and `immutable` => `const`), except that, as Manu pointed, the delegate could be changing the context so that breaks mutable => `const`.

If it applies to the delegate itself (`const` meaning "I don't alter the variables, just reading), then `const` can converts to 'mutable' but `immutable` can only convert to `shared const` AFAICS.

Reply:
> If you pass the delegate to a function, you simply can't enforce it

The function would need to have the appropriate qualifiers in its parameter type. This is how type checking already works: the body and the call site are in type checked according to the types of the parameters, and this tandem must ensure safety.

> Does the qualifier applies to the context, or the delegate itself ?

In `Q1(T delegate(A) Q2)`, where Q1 and Q2 are qualifiers, Q1 applies to the delegate and Q2 applies to the “_this_ argument to opCall” i.e. the context/environment just like with classes.

Reply:
> [...] delegate types should not be much different from single-method interfaces. [...]

I see what you mean now, thanks.

Going back to the examples you mentioned:
```D
const(void delegate()) f = delegate() {}; // ok
f(); // not ok: opCall does not accept const this
```
This is IMO problematic, as the RHS has a `const delegate` but the `const`ness is not inferred, so an untyped `delegate` is assigned to a `const` context variable.
If I understand correctly, this is roughly the same as doing:
```D
struct Bar { void foo (int) {} }
void main () { const Bar b; scope dg = &b.foo; }
```
Where we take the address of a mutable method with a `const` context. This is currently rejected by the compiler, although the error message isn't great:
```console
foo.d(2): Error: mutable method foo.Bar.foo is not callable using a const b
```
I think it would make sense to reject this, as it is nonsensical to have a `delegate` that cannot be called. There is only one instance of something similar in the language (`lazy`) but it doesn't seem to be intentional.

Likewise:
```D
const(void delegate() const) g = delegate() const {}; // ok
g(); // ok: opCall accepts const this
```
I assume that `(void delegate() const) g = delegate() const {};` is okay and callable as well ?

> In addition the following conversions should be possible:
```D
void delegate() const     -> const(void delegate() const)
void delegate() immutable -> immutable(void delegate() immutable)
void delegate() shared    -> shared(void delegate() shared)
```
Am I correct that the following should be possible as well ?
```D
void delegate() const     -> void delegate()
void delegate() immutable -> void delegate() [const] shared
```

Additionally, composition! In this approach, type ctors apply to the context, not the variable. We will experience the same usability issue as we have with classes right now, where there is no way to have a mutable reference to a `const` & co class, because `const Object o` is a `const` variable holding a reference to a `const Object`. I think this is fine though.

However, what happens in the following case:
```D
struct Bar { void delegate() var; }
void main ()
{
    const Bar b = { var: .... };
    pragma(msg, typeof(b.var));
}
```

Reply:
What's the risk of allowing `void delegate()` to implicitly convert to `void delegate() const` and vice versa? with `shared` there is the obvious synchronization hazard, but for `const` does it really matter?

Reply:
> What's the risk of allowing void delegate() to implicitly convert to void delegate() const

The `const` is applied to the `this` pointer of the delegate. A mutable pointer can always be implicitly converted to const.

> and vice versa?

A const pointer cannot be implicitly converted to mutable.

Reply:
But why does it matter here? The context pointer is an opaque pointer you can assume its size or structure thus you can't really modify it except in a very confined environment.
I am concerned because "fixing" this bug will disable passing random delegates to functions which receive something broad like a `void delegate()`.

Reply:
> The `const` is applied to the `this` pointer of the delegate. A mutable pointer can always be implicitly converted to const.

Yes but you convert the context pointer to `const` you have to replace the function pointer with a `const` function in order for this to be sound, but it isn't possible with implicit conversion.

Reply:
If you can implicitly convert a const delegate to a mutable delegate, then the .funcptr can be swapped with a mutating one, and change that supposedly const parameter.

Once a hole like this is opened up in the type system, all kinds of demons can find their way through it.

Reply:
This isn't as simple to solve as the slice case. The thing with a delegate is that it's a pair of opaque pointers, one for opaque data, and the other for an opaque instruction, you don't know where it originated from in order to swap the function pointer if there is even a const or mutable function.

So there are two options which look sound to me, 1) disable all implicit conventions related to storage class i.e `if (this.mod != to.mod) return MATCH.nomatch`, 2) keep things as they are with `const` and disable conversions with `shared`.

Reply:
@WalterBright @SSoulaimane A const delegate can be safely converted to a mutable delegate, but not vice-versa. Replacing the function pointer or the context pointer separately is not safe.